### PR TITLE
Working draft of SRAMCompiler

### PR DESCRIPTION
**Pull request by colinschmidt**
Ready for code review.
Works when consuming IR from https://github.com/ucb-bar/barstools/pull/50
See tech specific repo for an implmentation.

### CODE REVIEW

File: src/hammer-vlsi/defaults.yml [hunk 1]
```diff
@@ -35,6 +35,11 @@ vlsi.core:
   lvs_tool_path: ["${vlsi.builtins.hammer_vlsi_path}/lvs"]
   lvs_tool_path_meta: subst
 
+  # Path(s) to lvs tools.
```
Comment (edwardcwang) [inline src/hammer-vlsi/defaults.yml]: need to rename `lvs`

File: src/hammer-vlsi/defaults.yml [hunk 2]
```diff
@@ -35,6 +35,11 @@ vlsi.core:
...
+  # Add the built-in path to lvs tools.
+  gen_srams_tool_path: ["${vlsi.builtins.hammer_vlsi_path}/lvs"]
```
Comment (edwardcwang) [inline src/hammer-vlsi/defaults.yml]: lvs

File: src/hammer-vlsi/defaults.yml [hunk 3]
```diff
@@ -48,6 +53,9 @@ vlsi.core:
   # LVS tool to use.
   lvs_tool: "nop"
 
+  # SRAM Compiler tool to use.
+  gen_srams_tool: "nop"
```
Comment (edwardcwang) [inline src/hammer-vlsi/defaults.yml]: How about `sramcompiler_tool`, `sramgenerator_tool` or `sram_gen_tool`? (we'll need to rename the above as well)

File: src/hammer-vlsi/defaults.yml [hunk 4]
```diff
@@ -414,6 +425,13 @@ lvs.submit:
   settings: "vlsi.submit.settings"
   settings_meta: lazycrossref
 
+# inherit settings from vlsi.submit but allow us to override them
+gen_srams.submit:
```
Comment (edwardcwang) [inline src/hammer-vlsi/defaults.yml]: ditto above for name

File: src/hammer-vlsi/generate_properties.py [hunk 5]
```diff
@@ -153,6 +153,18 @@ def main(args) -> int:
                                         ]
                                         )
 
+    HammerSRAMCompilerTool = Interface(module="HammerSRAMCompilerTool",
```
Comment (edwardcwang) [inline src/hammer-vlsi/generate_properties.py]: I think we should try to align the name of the hammer IR name with the Python name as much as possible. In this case e.g. `sramcompiler_tool` <-> `SRAMCompilerTool`

File: src/hammer-vlsi/hammer_vlsi/cli_driver.py [hunk 6]
```diff
@@ -110,9 +110,14 @@ def __init__(self) -> None:
         self.par_rundir = ""  # type: Optional[str]
         self.drc_rundir = ""  # type: Optional[str]
         self.lvs_rundir = ""  # type: Optional[str]
+        self.gen_srams_rundir = ""  # type: Optional[str]
```
Comment (edwardcwang) [inline src/hammer-vlsi/hammer_vlsi/cli_driver.py]: ditto naming

File: src/hammer-vlsi/hammer_vlsi/constraints.py [hunk 7]
```diff
@@ -45,6 +45,31 @@ def from_setting(ilm: dict) -> "ILMStruct":
             lef=str(ilm["lef"])
         )
 
+#Transliterated-ish from SRAMGroup in MDF
```
Comment (edwardcwang) [inline src/hammer-vlsi/hammer_vlsi/constraints.py]: Let's add a TODO / file an issue to convert these to use the IR library #330

**Comment by edwardcwang** [comment]
How easy is it to add one unit test for this? I'm also OK if we file an issue, though things may break then :)

**Comment by colinschmidt** [comment]
I'll add a quick one.

**Pull request merged**